Shrink the size of ReportingObserver

rreno opened the pull request:
#### 12c97f7650721444b359ac0d0ab2e4d9b1599fda
<pre>
Shrink the size of ReportingObserver
<a href="https://bugs.webkit.org/show_bug.cgi?id=286232">https://bugs.webkit.org/show_bug.cgi?id=286232</a>
<a href="https://rdar.apple.com/143206158">rdar://143206158</a>

Reviewed by Anne van Kesteren.

In the spec ReportingObserverOptions contains a boolean and a
sequence&lt;DOMString&gt;. We had it implemented as an optional sequence which
took up extra space for std::optional bookkeeping. Before this patch
ReportingObserver was 104 bytes but with a 16 byte aligned allocator we
end up using 112 bytes. This shrinks the type down to 96 bytes by
storing a nullopt `types` array as an empty array in the ReportingObserver
C++ object.

* Source/WebCore/Modules/reporting/ReportingObserver.cpp:
(WebCore::ReportingObserver::ReportingObserver):
(WebCore::ReportingObserver::observe):
(WebCore::ReportingObserver::appendQueuedReportIfCorrectType):
* Source/WebCore/Modules/reporting/ReportingObserver.h:
* Source/WebCore/Modules/reporting/ReportingObserverOptions.idl: Removed.

Canonical link: <a href="https://commits.webkit.org/289153@main">https://commits.webkit.org/289153@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/75acd839f6f3783271dee33b25bf83fee9375cfc

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/85453 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/5171 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/39868 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/90565 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/36478 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/87525 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/5315 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/13149 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/66419 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-view-transitions/active-view-transition-type-on-non-root.html (failure)") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/24232 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/88483 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/4064 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/77595 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/46701 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/3951 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/31857 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/35548 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/74648 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/32704 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/92128 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/12786 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/9364 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/75065 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/13012 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/73428 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/74189 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18348 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/18500 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/16930 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/4857 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/12771 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/18197 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/12589 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/16049 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/14354 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

annevk commented on Source/WebCore/Modules/reporting/ReportingObserver.idl:
```diff
@@ -39,6 +39,6 @@ typedef sequence<Report> ReportList;
-    sequence<[AtomString] DOMString>? types;
+    sequence<[AtomString] DOMString> types;
     boolean buffered = false;
 };
```
Why do we define this dictionary twice?

rreno replied:
I'm not sure why. Let me experiment with getting rid of the other file.

rreno replied:
In fact the two files don't match as the IDL attribute `[AtomString]` is in one file but not the other.

rreno replied:
I deleted the standalone file and didn't see any ill behavior locally testing. Let's see what EWS says.

annevk replied:
Ah, nice! I wish we could find these kind of files more easily. I've seen that a few times now where a file that appears to be part of the build isn't actually part of it.

annevk commented on Source/WebCore/Modules/reporting/ReportingObserverOptions.idl:
```diff
@@ -26,6 +26,6 @@
 // https://w3c.github.io/reporting/#reportingobserver
 
 dictionary ReportingObserverOptions {
-    sequence<DOMString>? types;
```
This is web-observable. Now whatever takes `ReportingObserverOptions` will throw if `types` is null. This does appear to match the specification, but maybe there should be a test for it.

rreno replied:
`imported/w3c/web-platform-tests/reporting/bufferSize.html` has a constructor which takes a `ReportingObserverOptions` but types is not defined in the options object. It does not newly throw.

Looking at the generated code in `JSReportingObserver.cpp` I think this is fine. What ends up happening is the `Vector<AtomString>` in C++ is default-constructed to be a zero sized array.

```C++
// in convertDictionary<ReportingObserver::Options>
    if (!typesValue.isUndefined()) { // <-- HERE typesValue.isUndefined() is true when ReportingObserverOptions does not define a types property.
        auto typesConversionResult = convert<IDLSequence<IDLAtomStringAdaptor<IDLDOMString>>>(lexicalGlobalObject, typesValue);
        if (UNLIKELY(typesConversionResult.hasException(throwScope)))
            return ConversionResultException { };
        result.types = typesConversionResult.releaseReturnValue();
    }

```

rreno replied:
Oh you probably meant this?
```JavaScript
var observer = new ReportingObserver(() => { }, { types: null });
```

Yes, that newly throws. Hm. Since this array is no longer visible to the web after construction I wonder if I can contort the bindings generator to create an empty array if it is explicitly set to `null`.

rreno replied:
Actually, I fixed this by just storing the members of the options struct separately in `ReportingObserver` and converting from an `optional` to an empty array in the C++ constructor.

If I later discover this pattern happens often (optional type which can be represented by a smaller type in C++) maybe then it would be worth special casing the code generator.

annevk replied:
I think this is okay for now, but note that https://w3c.github.io/reporting/#reportingobserver says it's not nullable so eventually we do want to remove this type being nullable as you did initially (and add a WPT to make sure all browsers either already do that or converge on that).

rreno replied:
Yes that makes sense. The non-nullability is what drove the original change. However, Chrome implements `ReportingObserverOptions` as a std::optional as well and does not throw for the case from my comment above. I didn't want to hurt interoperability with this change.

A WPT is a good idea. My guess is none of the browsers on wpt.fyi pass. Though maybe it's a candidate for a spec issue in that case.

annevk replied:
Yeah, we could also get the spec updated I suppose. It seems rather pointless for this to be nullable, but if Chrome is not willing to change I guess we should do that.

rreno committed:
Shrink the size of ReportingObserver
https://bugs.webkit.org/show_bug.cgi?id=286232
rdar://143206158

Reviewed by Anne van Kesteren.

In the spec ReportingObserverOptions contains a boolean and a
sequence<DOMString>. We had it implemented as an optional sequence which
took up extra space for std::optional bookkeeping. Before this patch
ReportingObserver was 104 bytes but with a 16 byte aligned allocator we
end up using 112 bytes. This shrinks the type down to 96 bytes by
storing a nullopt `types` array as an empty array in the ReportingObserver
C++ object.

* Source/WebCore/Modules/reporting/ReportingObserver.cpp:
(WebCore::ReportingObserver::ReportingObserver):
(WebCore::ReportingObserver::observe):
(WebCore::ReportingObserver::appendQueuedReportIfCorrectType):
* Source/WebCore/Modules/reporting/ReportingObserver.h:
* Source/WebCore/Modules/reporting/ReportingObserverOptions.idl: Removed.

Canonical link: https://commits.webkit.org/289153@main
